### PR TITLE
Fix TypedArray initialization with another TypedArray

### LINKED ISSUE #3836: ICE: Assertion 'ecma_object_class_is (object_p, LIT_MAGIC_STRING_ARRAY_BUFFER_UL)' failed at jerry-core/ecma/operations/ecma-arraybuffer-object.c(ecma_arraybuffer_get_buffer):190. Error: ERR_FAILED_INTERNAL_ASSERTION
###### JerryScript revision

a56e31f194fe52000ed95605ae3c92390cd60ad5

###### Build platform
Ubuntu 18.04.4 LTS (Linux 4.15.0-91-generic x86_64)

###### Build steps

- first output

```sh
python tools/build.py --profile=es2015-subset --lto=off --compile-flag=-g \
--error-messages=on --debug --strip=off --logging=on \
 --compile-flag=-fsanitize=address --stack-limit=15
```

- second output

```sh
python tools/build.py --profile=es2015-subset --lto=off --compile-flag=-g  \
--strip=off --logging=on  --stack-limit=15
```

###### Test case

```
function main() {
var v3 = new Float64Array(6);
var v4 = v3.buffer;
v4.constructor = Uint8Array;
var v5 = new Float64Array(v3);
}
main();
```

###### Execution steps

```sh
ICE: Assertion 'ecma_object_class_is (object_p, LIT_MAGIC_STRING_ARRAY_BUFFER_UL)' failed at /home/jerryscript/jerry-core/ecma/operations/ecma-arraybuffer-object.c(ecma_arraybuffer_get_buffer):190.
Error: ERR_FAILED_INTERNAL_ASSERTION
Aborted (core dumped)
```

```
$ gdb -nx -q ~/jerryscript/build/bin/jerry -ex 'r 01.js'
Reading symbols from /home/jerryscript/build/bin/jerry...
Starting program: /home/jerryscript/build/bin/jerry 01.js

Program received signal SIGSEGV, Segmentation fault.
ecma_typedarray_create_object_with_typedarray (typedarray_id=ECMA_FLOAT64_ARRAY, element_size_shift=<optimized out>, proto_p=<optimized out>, typedarray_p=0x5555556bd408 <jerry_global_heap+480>)
    at /home/jerryscript/jerry-core/ecma/operations/ecma-typedarray-object.c:655
655	    memcpy (dst_buf_p, src_buf_p, array_length << element_size_shift);
(gdb) x/i $rip
=> 0x55555557654e <ecma_op_create_typedarray+346>:	rep movsb %ds:(%rsi),%es:(%rdi)
(gdb) i r rdi
rdi            0x3004100020008     844704103137288
```

### PR DESCRIPTION
**Pull request by szilagyiadam**
Fixes #3836

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 1]
```diff
@@ -534,15 +534,21 @@ ecma_typedarray_create_object_with_length (ecma_length_t array_length, /**< leng
 
     ecma_object_t *ctor_proto_p = ecma_get_object_from_value (ctor_proto);
 
-    ecma_value_t byte_length_val = ecma_make_uint32_value (byte_length);
-    ecma_value_t new_arraybuffer = ecma_op_function_construct (ctor_proto_p,
-                                                               ctor_proto_p,
-                                                               &byte_length_val,
-                                                               1);
+    ecma_object_t *data_p = ecma_op_get_prototype_from_constructor (ctor_proto_p,
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c:537]: How about simply `proto_p`?

File: tests/jerry/es2015/regression-test-issue-3836.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+function validate_typedarray (typedarray, result) {
+  assert(typedarray.length === result.length);
+  for (var i=0;i<typedarray.length;i++) {
```
Comment (galpeter) [inline tests/jerry/es2015/regression-test-issue-3836.js:17]: Missing spaces around `=`, `<`  characters and after the `;`.

File: tests/jerry/es2015/regression-test-issue-3836.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    assert(typedarray[i] === result[i]);
+  }
+}
+
+var v1 = new Float64Array(6);
+v1.buffer.constructor = Uint8Array;
+var v2 = new Float64Array(v1);
+
+assert(v2.buffer.constructor === Uint8Array);
+validate_typedarray(v2, [0,0,0,0,0,0]);
```
Comment (galpeter) [inline tests/jerry/es2015/regression-test-issue-3836.js:27]: missing spaces after the commas.

File: tests/jerry/es2015/regression-test-issue-3836.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+assert(v4.buffer.constructor === Float64Array);
+validate_typedarray(v4, [0,0,0,0,0,0]);
+
+var v5 = new Uint32Array(6);
+v5.buffer.constructor = Set;
+var v6 = new Uint8Array(v5);
+
+assert(v6.buffer.constructor === Set);
+validate_typedarray(v6, [0,0,0,0,0,0]);
```
Comment (galpeter) [inline tests/jerry/es2015/regression-test-issue-3836.js:41]: I propose to have an extra test case :
```js
var species_called = false;

var v3 = new Float64Array(6);
var v4 = v3.buffer;
v4.constructor = {
  get [Symbol.species] (){
    species_called = true;
    return Uint8Array;
  }
}
var v5 = new Float64Array(v3);

assert(species_called);
assert(Reflect.getPrototypeOf(v5.buffer) === Uint8Array.prototype);
```
That is the species symbol in this case should set the created array buffer's prototype to the desired typedarray's prototype.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 5]
```diff
@@ -534,15 +534,21 @@ ecma_typedarray_create_object_with_length (ecma_length_t array_length, /**< leng
 
     ecma_object_t *ctor_proto_p = ecma_get_object_from_value (ctor_proto);
 
-    ecma_value_t byte_length_val = ecma_make_uint32_value (byte_length);
-    ecma_value_t new_arraybuffer = ecma_op_function_construct (ctor_proto_p,
-                                                               ctor_proto_p,
-                                                               &byte_length_val,
-                                                               1);
+    ecma_object_t *prototype_p = ecma_op_get_prototype_from_constructor (ctor_proto_p,
+                                                                    ECMA_BUILTIN_ID_ARRAYBUFFER_PROTOTYPE);
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-typedarray-object.c:538]: Misaligned line.

**Pull request merged**